Rewrite of the content handling, both on client and on server.

sbordet opened the pull request:
Introduced a new io.Content class, with Content.Source, Content.Sink and Content.Chunk.

Updated the server-side to use io.Content instead of the previous server-side Content class.

Updated the client-side to use io.Content with a little specialization necessary for clients only.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Rewrite of the content handling, both on client and on server.

Introduced a new io.Content class, with Content.Source, Content.Sink and Content.Chunk.

Updated the server-side to use io.Content instead of the previous server-side Content class.

Updated the client-side to use io.Content with a little specialization necessary for clients only.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12.0.x-content'.

gregw commented:
@sbordet I think some of those test failures are me.  I must have broken the build and then gone on vacation 2 days.... how much into the jar for that?
Fixing now and will tell you when you can merge

sbordet commented:
@gregw we found the issue and pushed a fix to `jetty-12.0.x`.

joakime commented:
> @gregw we found the issue and pushed a fix to `jetty-12.0.x`.

@sbordet looks like we broke the ee10 side with that fix (commit bfcd7845eb43f0c3d73b2e0f2369eff306cdb58b)

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -0,0 +1,403 @@
+/**
+ * TODO: expand javadocs.
+ */
+public interface Content
```
Since Content is only a name scope, should it be a class with a private constructor_

sbordet replied:
It's not only a name scope. It has `copy()` and possibly in future some other utility methods that take both `Source` and `Sink`.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
Javadoc doesn't need to be complete, but things likethis should be fixed

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
How about `isComplete`? That has a more transactional feel to it. 
Or `isContentComplete`?        Eitherway, pick a better name and remove the TODO.  If we kept TODOs for all the sub-optimal names, then the code would just be TODOs

sbordet replied:
`isTerminal()` is as transactional as `isComplete()` to me, so I'd just remove the TODO?

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
Strange that this is a nested class but `ByteBufferChunk` is not.

sbordet replied:
Because `ByteBufferChunk` does not have extra APIs other than those of `Chunk`. `Chunk.Error` has `getCause()`.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```suggestion

￼
￼        public static Chunk from(Throwable error)
￼        {
￼            return new Error(error);
￼        }
```

sbordet replied:
done.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ByteBufferChunk.java:
```diff
@@ -0,0 +1,83 @@
+
+    public void release()
+    {
+        if (releaser != null)
+        {
+            Runnable runnable = releaser.getAndSet(null);
+            if (runnable != null)
+                runnable.run();
+        }
```
would it not be better for the releaser to be a `Consumer<ByteBuffer>`?

sbordet replied:
No, for tests we typically use `CountDownLatch::countDown`, and for other cases we don't use the chunk's ByteBuffer because we delegate the call to something else, e.g. a RetainableByteBuffer, etc.
I think Runnable is better.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ContentOLD.java:
```diff
@@ -45,7 +46,7 @@
  * included in the read/write APIs.
  */
-public interface Content
+public interface ContentOLD
```
delete this

sbordet replied:
Ah yes.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -0,0 +1,403 @@
+     * }
+     * }</pre>
+     */
+    public interface Source
```
Whilst I do like the symmetry of `Content.Source` and `Content.Sink` it does introduce a name only interface.
Whilst I like the name `Content` it is ultimately doing nothing for us here and violates least surprise when you can't do `content.getLength()` and instead have to ask the "source" of your content for the content length.  Then the "source" of the "content" doesn't provide content, it provides chunks.

Alternates are:

 + leave as is
 + move chunk API to top, so we have `Chunk.Source` and `Chunk.Sink`
 + make `Content` the source API so we just have `Content` and `Content.Sink`

sbordet replied:
See above, `Content` is not a namespace only interface.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
Do we need an `Flow.Subscriber<Chunk> asSubscriber(Sink sink)` ?

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -165,37 +167,38 @@ public interface Request extends Attributes, Content.Reader
      *
-     * @see #demandContent(Runnable)
+     * @see #demand(Runnable)
      */
```
should remove javadoc here and let interface fully define semantics

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -165,37 +167,38 @@ public interface Request extends Attributes, Content.Reader
      *
-     * @see #readContent()
+     * @see #read()
      */
```
remove javadoc here

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -469,7 +461,7 @@ public void reset()
     {
-        return Content.asOutputStream(response);
+        return Content.Sink.asOutputStream(response);
     }
```
Do we need this convenience method?

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -57,18 +57,10 @@ public interface Response extends Content.Writer
-    {
-        write(last, callback, StandardCharsets.UTF_8.encode(utf8Content));
-    }
+    public void write(boolean last, Callback callback, ByteBuffer... buffers);
```
Do we need this empty override?  At very least do:
```suggestion
    void write(boolean last, Callback callback, ByteBuffer... buffers);
```

sbordet replied:
Done.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1073,6 +1079,22 @@ private HttpFields takeTrailers()
         }
 
+        @Override
+        public void write(Content.Chunk chunk, Callback callback)
```
might need to define this method in the interface so we can add javadoc saying that Trailers are handled (and what steps have to be done to get them)

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1073,6 +1079,22 @@ private HttpFields takeTrailers()
+                ResponseHttpFields trailers = _request.getHttpChannel()._responseTrailers;
+                if (trailers != null)
+                    trailers.add(trailersChunk.getTrailers());
+                callback.succeeded();
```
Don't we still need to write here?
```suggestion
                Response.super.write(true, callback);
```

sbordet replied:
Done.

gregw reviewed (changes requested):
just some initial feedback

lorban commented on jetty-core/jetty-fcgi/fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/internal/HttpStreamOverFCGI.java:
```diff
@@ -187,14 +187,18 @@ private void notifyContentAvailable()
-    public void onContent(Content content)
+    public void onContent(Content.Chunk chunk)
     {
-        _content = content;
```
Maybe you could add a paranoid check or an assertion here?
```java
assert _chunk == null;
```

lorban commented on jetty-core/jetty-fcgi/fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/internal/HttpStreamOverFCGI.java:
```diff
@@ -187,14 +187,18 @@ private void notifyContentAvailable()
-        _content = Content.last(_content);
+        if (_chunk == null)
+            _chunk = Content.Chunk.EOF;
+        // TODO: paranoid check, maybe remove it.
```
I quite like those paranoid checks, they catch bugs earlier and are cheap to compute. I'd keep it or at least turn it into an assertion.

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/util/ByteBufferRequestContent.java:
```diff
@@ -13,77 +13,47 @@
-    }
-
-    private class SubscriptionImpl extends AbstractSubscription
+    public boolean rewind()
```
This override is useless as it only calls super.

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/util/MultiPartRequestContent.java:
```diff
@@ -54,7 +53,7 @@
  * </pre>
  */
-public class MultiPartRequestContent extends AbstractRequestContent implements Closeable
+public class MultiPartRequestContent /*extends AbstractRequestContent*/ implements Request.Content, Closeable
```
This class looks unfinished. Did you leave it as-is on purpose?

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/util/PathRequestContent.java:
```diff
@@ -66,107 +49,20 @@ public PathRequestContent(String contentType, Path filePath) throws IOException
-    }
-
-    public void setByteBufferPool(ByteBufferPool byteBufferPool)
+    public boolean rewind()
```
This override is unneeded as it only calls super.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ContentProducer.java:
```diff
@@ -104,16 +104,17 @@
      * and has no available content.
      */
-    Content nextContent();
+    Content.Chunk nextContent();
```
This should be renamed `nextChunk()` or maybe just `next()`.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ContentProducer.java:
```diff
@@ -89,11 +89,11 @@
     boolean hasContent();
```
Now that this class isn't dealing with `Content`s but `Chunk`s this method should be renamed. Maybe to `hasChunk()`?

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ContentProducer.java:
```diff
@@ -59,17 +59,17 @@
-     * @return the byte count produced by the underlying {@link org.eclipse.jetty.server.Content.Reader}.
+     * @return the byte count produced by the underlying {@link Content.Source}.
      */
     long getRawContentArrived();
```
Now that this class isn't dealing with `Content`s but `Chunk`s this method should be renamed.  Maybe to `getRawBytesArrived()`?

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ContentProducer.java:
```diff
@@ -13,12 +13,12 @@
- * ContentProducer is the bridge between {@link HttpInput} and {@link org.eclipse.jetty.server.Content.Reader}.
+ * ContentProducer is the bridge between {@link HttpInput} and {@link Content.Source}.
  */
 public interface ContentProducer
```
Now that this class isn't dealing with `Content`s but `Chunk`s this class should be renamed.

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12.0.x-content'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ContentProducer.java:
```diff
@@ -104,16 +104,17 @@
      * and updating this instance' internal state.
      */
-    void reclaim(Content content);
+    void reclaim(Content.Chunk content);
```
Now that this class isn't dealing with `Content`s but `Chunk`s the following method `onContentProducible` should be renamed.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpStream.java:
```diff
@@ -19,20 +19,20 @@
     long getNanoTimeStamp();
 
-    Content readContent();
+    Content.Chunk readContent();
```
Now that this class is dealing with `Chunk`s instead of `Content`s this method should be renamed.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpStream.java:
```diff
@@ -19,20 +19,20 @@
-    Content readContent();
+    Content.Chunk readContent();
 
     void demandContent(); // Calls back on Channel#onDataAvailable
```
Now that this class is dealing with `Chunk`s instead of `Content`s this method should be renamed. And the comment points to an invalid method name that should be fixed too. :-)

lorban reviewed (changes requested):
First set of review comments, more will follow.

sbordet committed:
Fixes after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12.0.x-content'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed compilation issue.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -177,16 +177,16 @@ public void checkMinDataRate()
         if (LOG.isDebugEnabled())
-            LOG.debug("getRawContentArrived = {} {}", _rawContentArrived, this);
-        return _rawContentArrived;
+            LOG.debug("getRawContentArrived = {} {}", _rawBytesArrived, this);
```
The debug message's method name should be changed.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -207,38 +207,38 @@ public boolean consumeAll()
         return atEof;
     }
 
     private void failCurrentContent(Throwable x)
```
This method is actually failing the current chunk.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -251,10 +251,10 @@ public boolean onContentProducible()
-        Content content = nextTransformedContent();
+        Content.Chunk content = nextTransformedContent();
         if (LOG.isDebugEnabled())
             LOG.debug("nextContent = {} {}", content, this);
```
The debug message's method name should be changed.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -306,52 +306,52 @@ boolean isUnready()
     }
 
-    private Content nextTransformedContent()
+    private Content.Chunk nextTransformedContent()
```
This method is actually returning the next transformed chunk.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -306,52 +306,52 @@ boolean isUnready()
+    private Content.Chunk nextTransformedContent()
     {
         if (LOG.isDebugEnabled())
             LOG.debug("nextTransformedContent {}", this);
```
The debug message's method name should be changed if the method gets renamed.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -476,20 +476,20 @@ else if (content != _rawContent && !_rawContent.isSpecial() && !_rawContent.isEm
     }
 
-    private Content produceRawContent()
+    private Content.Chunk produceRawContent()
```
The method actually produces a raw chunk.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -476,20 +476,20 @@ else if (content != _rawContent && !_rawContent.isSpecial() && !_rawContent.isEm
                 _firstByteTimeStamp = System.nanoTime();
             if (LOG.isDebugEnabled())
-                LOG.debug("produceRawContent updated rawContentArrived to {} and firstByteTimeStamp to {} {}", _rawContentArrived, _firstByteTimeStamp, this);
+                LOG.debug("produceRawContent updated rawContentArrived to {} and firstByteTimeStamp to {} {}", _rawBytesArrived, _firstByteTimeStamp, this);
```
The debug message's `rawContentArrived` should be renamed, as well as the method name should be changed if the method gets renamed.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -476,20 +476,20 @@ else if (content != _rawContent && !_rawContent.isSpecial() && !_rawContent.isEm
+                LOG.debug("produceRawContent updated rawContentArrived to {} and firstByteTimeStamp to {} {}", _rawBytesArrived, _firstByteTimeStamp, this);
         }
         if (LOG.isDebugEnabled())
             LOG.debug("produceRawContent produced {} {}", content, this);
```
The debug message's method name should be changed if the method gets renamed.

lorban commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpInput.java:
```diff
@@ -155,17 +155,17 @@ public long getContentReceived()
         try (AutoLock lock = _contentProducer.lock())
         {
             if (LOG.isDebugEnabled())
                 LOG.debug("consumeAll {}", this);
```
The method name should be changed in the debug message.

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/AsyncContentProducer.java:
```diff
@@ -33,13 +34,13 @@
 class AsyncContentProducer implements ContentProducer
```
The renames made in the ee10 module are also valid here. I suppose they weren't made on purpose?

lorban commented on jetty-core/jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/internal/HttpStreamOverHTTP2.java:
```diff
@@ -163,22 +164,9 @@ public Runnable onData(DataFrame frame, Callback callback)
     }
 
-    private Content.Abstract newContent(DataFrame frame, Runnable complete)
+    private Content.Chunk newChunk(DataFrame frame, Runnable complete)
```
This method can probably be inlined now.

lorban commented on jetty-core/jetty-http3/http3-server/src/main/java/org/eclipse/jetty/http3/server/internal/HttpStreamOverHTTP3.java:
```diff
@@ -158,26 +159,13 @@ public Runnable onDataAvailable()
     }
 
-    private Content.Abstract newContent(Stream.Data data)
+    private Content.Chunk newChunk(Stream.Data data)
```
This method can probably be inlined now.

lorban commented on jetty-ee10/jetty-ee10-proxy/src/test/java/org/eclipse/jetty/ee10/proxy/AsyncMiddleManServletTest.java:
```diff
@@ -66,10 +66,10 @@
 import org.eclipse.jetty.server.HttpConnectionFactory;
 import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.server.ServerConnector;
-//import org.eclipse.jetty.server.internal.HttpChannelState;
```
Can be removed.

lorban commented on jetty-ee10/jetty-ee10-proxy/src/test/java/org/eclipse/jetty/ee10/proxy/ProxyServletFailureTest.java:
```diff
@@ -40,11 +40,9 @@
 import org.eclipse.jetty.server.HttpConnectionFactory;
 import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.server.ServerConnector;
-//import org.eclipse.jetty.server.internal.HttpChannelState;
```
Can be removed.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpStream.java:
```diff
@@ -19,22 +19,36 @@
+     * <p>Demands more content chunks to the underlying implementation.</p>
+     * <p>This method is called from the implementation of {@link Request#demand(Runnable)}.</p>
+     */
+    void demand();
```
Don't we want to document what gets called back?

lorban reviewed (changes requested):
A few minor things left, but LGTM.

sbordet committed:
Fixes after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12.0.x-content'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Made Response.write(Chunk, Callback) abstract.
The reason is that when Response is wrapped, both write(Chunk, Callback) and write(boolean, Callback, ByteBuffer...) must be overridden otherwise
one method operates in the outer instance, while the other method operates in the inner instance.

For the same reason, Response.write(boolean, Callback, String) was made static.

Updated the code that was impacted by the above changes.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

joakime commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java:
```diff
@@ -51,9 +51,6 @@ protected Action process() throws Throwable
         }
 
-        if (current instanceof Content.Chunk.Error error)
-            throw error.getCause();
```
Is Content.Chunk.Error handled somewhere else now?

sbordet replied:
Yes. Implementations of `Sink.write(Chunk, Callback)` will fail the callback if they receive a `Chunk.Error`.

sbordet committed:
Fixed checkstyle.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSinkSubscriber.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    public void onError(Throwable throwable)
+    {
+        sink.write(Content.Chunk.from(throwable), Callback.NOOP);
```
This implementation should actually result in a no-op since the `write(Chunk)` implementation is supposed to call the callback's `fail` method when the chunk is an error.

I'm not sure what to do in this case.

lorban reviewed (approved):
LGTM despite the suspicious `Flow` impl.

sbordet committed:
Removed Content.Sink.write(Chunk, Callback) because it forces every subclass to override both write(Chunk, Callback) and write(boolean, Callback, ByteBuffer...).

The only reason to have write(Chunk, Callback) was to be able to write Chunk.Error, or Trailers, but that is now supported in Content.copy() via Response.writeTrailers().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
@lorban Can you expand a bit here on the discussion you referenced in the email regarding `write(Chunk)`?
Specifically exactly why it is bad to have both.

I guess it makes the job of somebody that wants to wrap the response harder as they wrap write(boolean,ByteBuffer) and then themselves get wrapped, then either:
 1) the wrapper doesn't implement `write(Chunk)` and thus any ability to write trailers is lost, but at least the wrapped write is called; OR
 2) the wrapper does implement `write(Chunk)`, in which case the wrapped write is bypassed.

Ultimately I see `write(Chunk)` as 90% convenience method equivalent to `write(chunk.isLast(), chunk.getByteBuffer())` and 10% method for sending trailers....
So perhaps we just have a static `write(Chunk)` that does both the convenience thing, but also knows how to handle trailers.   We then always know that `write(Chunk)` will be on the outermost wrapper and both the trailers API and write API may be individually intercepted with specific APIs

gregw commented:
@lorban The other question about the write side is should it be `write(boolean, Callback, ByteBuffer...)` or `write(boolean, Callback, ByteBuffer)`?
The former is very rarely used, if ever.  But it costs a bit of allocation and complexity.      However it does have nice flush semantics of passing no buffers at all.
The later will need an empty buffer passed for flushes, but is simpler a implementation and covers 99.9999999% of the use-cases.

It is different at the lower layers, were we are often dealing with headers and body, thus gather writes make sense and can go all the way down to the network interface.... but for an application on top of http/2 or http/3, there is no way the gather write is going all the way down.  I guess it might influence the framing of the http>=2, but  is the cost of the gather worth it?     It may well be better to aggregate rather than gather.

lorban commented:
@gregw you mostly figured the problem: having more than one `write()` method forces you to be careful about how and when to call the wrapped Response. But it's more subtle than that: `write(Chunk, cb)` has to handle error content (which can be done at the io layer) and trailers content (which cannot be done at the io layer) so it's tempting to give it a default implementation that almost always does the wrong thing when Response is implemented.

This is why in the current PR's state `Content.write(Chunk, cb)` is overridden in Response to remove the default impl. And you also have the oddity of AsyncContent implementing `write(last, cb, ByteBuffer...)` by calling `write(Chunk, cb)`. During the multiple reviews of the code I discussed with @sbordet we figured that more often than not the principle of least surprise was shattered and there too often were non-obvious things to keep in mind w.r.t the implementation of those two `write()`.

Personally, I'm not fond of the `write(boolean, Callback, ByteBuffer...)` signature as being in a situation where we can perform gather writes IMHO is more of an exceptional than a common use-case. But I've also expressed my preference for `write(Chunk, cb)` which has the downside of forcing an allocation in some cases, but not in the most common one. `write(boolean, Callback, ByteBuffer)` would still be an improvement and I don't think having to pass an empty buffer instead of no buffer to signal a flush is a big deal.

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-content

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

# Conflicts:
#	jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/gzip/GzipHandlerTest.java
#	jetty-ee10/jetty-ee10-glassfish-jstl/src/test/webapp/WEB-INF/lib/testtaglib.jar

lorban commented:
@gregw maybe we should get this merged before discussing the `write(xxx)` signatures? I think the current state of this PR is reasonable enough to get merged as-is and eventually reworking `write(xxx)` later doesn't sound too much of a big deal.

gregw commented:
@lorban @sbordet I tend to agree that we should merge this prior to resolving the write signature issue.    There is enough goodness here that we don't want it to go stale and maintaining 2 dev branches is needless double work.

Note also that the decision between `write(boolean, Callback, ByteBuffer)` and `write(Chunk, cb)` needs to be informed by benchmarks.  I too like the symmetry of the later and with `Chunk.of(buffer)` style coding it is simple enough to use.    It might also not be that much extra allocation for some common cases, as servlet HttpOutput will have an aggregating buffer, which could have a preallocated Chunk (or HttpOutput could be the Chunk itself).

It may even be that if we keep `write(boolean, Callback, ByteBuffer)` as a default wrapper of `write(Chunk.from(last,buffer), cb)` then the two APIs do work OK side by side?   Hmm maybe now I'm just talking myself into keeping both.... let's merge and think about it.

Build is green so let's go!

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
This is actually wrong because it ignores the release conversation. Needs to be:
```suggestion
            if (chunk instanceof Content.Chunk.Error error)
            {
                chunk.release();
                callback.failed(error.getCause());
            }
            else
                write(chunk.isLast(), Callback.from(chunk::release, callback), chunk.getByteBuffer());
```

lorban replied:
For the happy path, that's _another_ allocation on the fast path.
For the error path, since that `release()` call is a no-op I wonder if we should not document it as such on `Chunk.Error`.

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12.0.x-content'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed imports.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed failing tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Changed signature of Sink.write() to take a single buffer.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Changed signature of HttpStream.send() to take a single buffer.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -0,0 +1,628 @@
+     * @param callback the callback to notify when the copy is complete
+     * @param chunkHandler an optional predicate to handle the current chunk and its callback
+     */
+    public static void copy(Source source, Sink sink, Callback callback, BiPredicate<Chunk, Callback> chunkHandler)
```
maybe make the callback the last arg to match all other APIs with the callback last.

gregw replied:
Maybe have a version like:
```suggestion
    public static void copy(Source source, Sink sink, Callback callback)
```

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -0,0 +1,628 @@
+     * @param source the source to copy from
+     * @param sink the sink to copy to
+     * @param callback the callback to notify when the copy is complete
+     * @param chunkHandler an optional predicate to handle the current chunk and its callback
```
or null

gregw reviewed (approved):
Other than some niggles +1

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12.0.x-content'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Made Callback the last parameter in Content.copy().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'jetty-12.0.x-content'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
